Sources/ParseSwift/API/API+Commands.swift

```diff
@@ -34,27 +35,60 @@ internal extension API {
         }
 
         public func execute(options: API.Options) throws -> U {
+            var responseResult: Result<U, ParseError>?
+
+            let group = DispatchGroup()
+            group.enter()
+            self.executeAsync(options: options, callbackQueue: nil) { result in
+                responseResult = result
+                group.leave()
+            }
+            group.wait()
+
+            guard let response = responseResult else {
+                throw ParseError(code: .unknownError,
+                                 message: "couldn't unrwrap server response")
+            }
+            return try response.get()
+        }
+
+        public func executeAsync(options: API.Options, callbackQueue: DispatchQueue?,
+                                 completion: @escaping(Result<U, ParseError>) -> Void) {
             let params = self.params?.getQueryItems()
             let headers = API.getHeaders(options: options)
             let url = ParseConfiguration.serverURL.appendingPathComponent(path.urlComponent)
 
-            var components = URLComponents(url: url, resolvingAgainstBaseURL: false)!
+            guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+                let urlComponents = components.url else {
+                    completion(.failure(ParseError(code: .unknownError,
+                                                   message: "couldn't unrwrap url components for \(url)")))
+                return
+            }
             components.queryItems = params
 
-            var urlRequest = URLRequest(url: components.url!)
+            var urlRequest = URLRequest(url: urlComponents)
             urlRequest.allHTTPHeaderFields = headers
             if let body = data {
                 urlRequest.httpBody = body
             }
             urlRequest.httpMethod = method.rawValue
-            let responseData = try URLSession.shared.syncDataTask(with: urlRequest).get()
-            do {
-                return try mapper(responseData)
-            } catch {
-                do {
-                    throw try getDecoder().decode(ParseError.self, from: responseData)
-                } catch {
-                    throw ParseError(code: .unknownError, message: "cannot decode response: \(error)")
+
+            URLSession.shared.dataTask(with: urlRequest, callbackQueue: callbackQueue) { result in
+                switch result {
+
+                case .success(let responseData):
```

Comment:
What do you think the other way should be?

I think there’s two types of success, one is just a standard HTTP(s) success which is just being able to connect and the other is communicating to the parse server correctly

Reply:
That makes sense; what if HTTP failures AND Parse errors both fall under the failure case then?

Reply:
This is a good question...

For some reason github isn’t let me post on the lines directly, but if the parse-server sends an error it will hit the completion in line 88. Any other error will hit the completion on line 85. I made sure to test parse-errors, http errors, and a JSON error in the test cases

Reply:
Can you take a look at commit 07a89ca on the `async_api` branch in my fork? I updated `URLSession.dataTask` to take the mapper as one of its parameters; I think it reads better, and that logic might be more appropriate for the `URLSession` extension anyway.

Reply:
I like it, I added your changes in